acceptance: TestDockerCLI_test_demo_partitioning failed [slow quiesce in `rpc.(*peer).launch`]

acceptance.TestDockerCLI_test_demo_partitioning [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18036192?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18036192?buildTab=artifacts#/) on master @ [74e821f1d1ce358f49354f8581b8663aa4c2fb15](https://github.com/cockroachdb/cockroach/commits/74e821f1d1ce358f49354f8581b8663aa4c2fb15):


```
=== RUN   TestDockerCLI_test_demo_partitioning
    test_log_scope.go:165: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/b38fdc3e3d87d99024418e1ae9af5f4c/logTestDockerCLI_test_demo_partitioning1199483594
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:147: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/b38fdc3e3d87d99024418e1ae9af5f4c/logTestDockerCLI_test_demo_partitioning1199483594
--- FAIL: TestDockerCLI_test_demo_partitioning (368.22s)
=== RUN   TestDockerCLI_test_demo_partitioning/test_demo_partitioning.tcl
    util_docker.go:255: non-zero exit code: 1
    --- FAIL: TestDockerCLI_test_demo_partitioning/test_demo_partitioning.tcl (361.86s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_demo_partitioning.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45224

acceptance.TestDockerCLI_test_demo_partitioning [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18036712?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18036712?buildTab=artifacts#/) on master @ [5a7850a72f941992b1bb4b23a73b5fa5e9f15a68](https://github.com/cockroachdb/cockroach/commits/5a7850a72f941992b1bb4b23a73b5fa5e9f15a68):


```
=== RUN   TestDockerCLI_test_demo_partitioning
    test_log_scope.go:165: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/b38fdc3e3d87d99024418e1ae9af5f4c/logTestDockerCLI_test_demo_partitioning391537483
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:147: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/b38fdc3e3d87d99024418e1ae9af5f4c/logTestDockerCLI_test_demo_partitioning391537483
--- FAIL: TestDockerCLI_test_demo_partitioning (374.66s)
=== RUN   TestDockerCLI_test_demo_partitioning/test_demo_partitioning.tcl
    util_docker.go:255: non-zero exit code: 1
    --- FAIL: TestDockerCLI_test_demo_partitioning/test_demo_partitioning.tcl (365.84s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_demo_partitioning.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go
+++ b/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go
@@ -41,6 +41,14 @@ import (
 
 const elemName = "somestring"
 
+var setTelemetryHttpTimeout = func(newVal time.Duration) func() {
+	prior := diagnostics.TelemetryHttpTimeout
+	diagnostics.TelemetryHttpTimeout = newVal
+	return func() {
+		diagnostics.TelemetryHttpTimeout = prior
+	}
+}
+
 func TestTenantReport(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -297,7 +305,7 @@ func TestTelemetry_SuccessfulTelemetryPing(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	diagnostics.TelemetryHttpTimeout = 3 * time.Second
+	defer setTelemetryHttpTimeout(3 * time.Second)()
 	rt := startReporterTest(t, base.TestIsSpecificToStorageLayerAndNeedsASystemTenant)
 	defer rt.Close()
 
@@ -364,6 +372,29 @@ func TestTelemetry_SuccessfulTelemetryPing(t *testing.T) {
 		})
 	}
 
+}
+
+// This test will block on `stopper.Stop` if the diagnostics reporter
+// doesn't honor stopper quiescence when making its HTTP request.
+func TestTelemetryQuiesce(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	defer setTelemetryHttpTimeout(10 * time.Minute)()
+	rt := startReporterTest(t, base.TestIsSpecificToStorageLayerAndNeedsASystemTenant)
+	defer rt.Close()
+
+	ctx := context.Background()
+	setupCluster(t, rt.serverDB)
+
+	// Ensure that we block for long enough to trigger test timeout.
+	defer rt.diagServer.SetWaitSeconds(15 * 60)()
+	dr := rt.server.DiagnosticsReporter().(*diagnostics.Reporter)
+	stopper := rt.server.Stopper()
+
+	dr.PeriodicallyReportDiagnostics(ctx, stopper)
+	stopper.Stop(ctx)
+	<-stopper.IsStopped()
 }
 
 func TestUsageQuantization(t *testing.T) {

--- a/pkg/server/diagnostics/reporter.go
+++ b/pkg/server/diagnostics/reporter.go
@@ -192,7 +192,11 @@ func (r *Reporter) PeriodicallyReportDiagnostics(ctx context.Context, stopper *s
 			// Consider something like rand.Float() > resetFreq/reportFreq here to sample
 			// stat reset periods for reporting.
 			if shouldReportDiagnostics(ctx, r.Settings) {
-				r.ReportDiagnostics(ctx)
+				func() {
+					ctx, cancel := stopper.WithCancelOnQuiesce(ctx)
+					defer cancel()
+					r.ReportDiagnostics(ctx)
+				}()
 			}
 
 			nextReport = nextReport.Add(reportFrequency.Get(&r.Settings.SV))

